### ISSUE TITLE
Convert JSON or Dict to client objects

### ISSUE THREAD

**Issue by remram44** [kubernetes-client/python#977]
See also https://github.com/kubernetes-client/python/issues/340#issuecomment-526249075 (was closed for inactivity, NOT resolved)
See also #63

**Motivation** 
It is sometimes useful to be able to load Python objects from JSON, for example to read it from a config file. It should be possible to get Python objects from JSON, to pass around the code in a type-safe manner, for modification and ultimately sending through the client.

**Feature request** 
*There should be functions to convert JSON to objects and objects to JSON. This code already exists (it's required to talk to the API server), it should be exposed.*

**Workaround?** 
Right now, you can put this raw JSON into Python objects' fields, and the API client will apparently accept this, however if you use such mixed objects your code will probably break since their attribute names are different (API uses PascalCase, Python objects use snake_case). Even if taking care to handle that, you will reach the point where you have to handle things coming from the API client (Python objects using snake_case) differently from what you load from config files (JSON using PascalCase). Really messy.

Right now I am left considering dropping this lib altogether and issuing requests directly, since the point of the wrapper is to expose safe typed Python classes instead of the raw JSON, but this doesn't work if you use any kind of config file.

**Current status** 
It is somewhat possible to convert from a Python object to JSON using `client.api_client.sanitize_for_serialization()`, however converting from JSON to Python is completely impossible; the `deserialize()` method that exists expects a `requests.Response`, not JSON. Calling `client.api_client._ApiClient__deserialize()` works, but that is a private method, and you need to pass it the `kind` read from the JSON yourself.

Apologies for duplicating #340 but I cannot re-open it myself.

**Comment by chekolyn** [comment]
I was just looking trying to do this. This would be a useful function.

**Comment by yliaog** [comment]
@chekolyn could you take this issue, and work on a pr?

**Comment by chekolyn** [comment]
@yliaog sure, I'll give it a shot. i think the deserializer() might have been auto generated by swagger codegen. So where would this be appropriate? utils?

**Comment by micw523** [comment]
Yes that looks like generated code. I think utils should be fine, but somewhere in the base repo may work as well since we don’t necessarily require e2e tests on this one (do we?)

**Comment by chekolyn** [comment]
I've been doing some tests and using a similar approach to [utils.create_from_yaml ](https://github.com/kubernetes-client/python/blob/master/kubernetes/utils/create_from_yaml.py)to infer the object  response_type to use in the already generated codegen deserialize function.

Using this approach is possible to deserialize single objects and a list of multiple items to a list of  individual namespaced objects; however this not covered all the different deserialize options we have. 
For example we have:
- list_xxx_for_all_namespaces
- list_namespaced_xxx

My question is is there a need to deserialize to these type of objects? Or deserializing to single objects covers the use case? One complication of this is when the is a List with multiple kind of objects (something like json output of` kubectl get all -o json`)

@remram44 will the proposed solution work for you? 
Anyone else has any opinions or suggestions ?

**Comment by remram44** [comment]
I'm not sure what you mean, lists what kind of lists wouldn't be deserializable?

**Comment by chekolyn** [comment]
Is really more about the object we output and not the input.
And almost every object has a list object as well.
 
For example:
pods = core_v1.list_pod_for_all_namespaces()
type(pods)
kubernetes.client.models.v1_pod_list.V1PodList

type(pods.items[0])
kubernetes.client.models.v1_pod.V1Pod

So in the proposed solution we would output a native python list/array that has individually deserialize objects. 

Something like this:
```
objs = deserialize.load_from_json('/Users/user/all.ns.data.json')

for obj in objs:
  print(type(obj))
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_pod.V1Pod'>
<class 'kubernetes.client.models.v1_service.V1Service'>
<class 'kubernetes.client.models.v1_service.V1Service'>
<class 'kubernetes.client.models.v1_service.V1Service'>
<class 'kubernetes.client.models.v1_service.V1Service'>
<class 'kubernetes.client.models.v1_daemon_set.V1DaemonSet'>
<class 'kubernetes.client.models.v1_deployment.V1Deployment'>
<class 'kubernetes.client.models.v1_deployment.V1Deployment'>
<class 'kubernetes.client.models.v1_replica_set.V1ReplicaSet'>
<class 'kubernetes.client.models.v1_replica_set.V1ReplicaSet'>
<class 'kubernetes.client.models.v1_replica_set.V1ReplicaSet'>

type(objs)
list
```

The alternative is to separate the list items by kind and return another 
native python list but instead of individual objects return k8s list objects that would be accesible via the each obj.items
```
kubernetes.client.models.v1_pod_list.V1PodList
kubernetes.client.models.v1_service_list.V1ServiceList
etc
```
The reason I mentioned this is that they way we would iterate over the items would be different. 
I'm inclined to take the first approach since it's less complicated, but was wondering if anyone needs to deserialize to these other k8s List objects.

**Comment by davlum** [comment]
I see there is a PR already open for this functionality. I'm not sure if the discussion is better had here or on the PR directly. There has been talk over on the [apache-airflow](https://github.com/apache/airflow) repo of how useful the [private deserialize methods would be exposed as public](https://github.com/apache/airflow/pull/6230#discussion_r336869991). It seems to me that the way the private method approaches this is the simplest, leaving it to the caller to also pass in the `kind`. Are there not ambiguities when trying to infer the `kind`? For example if I have a json `{ "name": "source" }` this could be inferred as a `V1SecretEnvSource` or a `V1ConfigMapEnvSource`

**Comment by chekolyn** [comment]
@davlum if I understand correctly, you would like to pass the object type directly; perhaps we can pass it along and not infer it, basically skip this line: https://github.com/chekolyn/python/blob/json_deserialize/kubernetes/utils/deserialize.py#L252
and pass the value directly. 

Additionally, we look for more than just the kind to infer an object. We look for the api version and api group.

Do you have any json/yaml examples I can test with?

**Comment by davlum** [comment]
Yeah, I'm essentially looking for the functionality of [`__deserialize_model(self, data, klass):`](https://github.com/kubernetes-client/python/blob/ca4f31198e9748f1075af0e02abb8f7e1e174159/kubernetes/client/api_client.py#L614) but made public. Because you have to pass the model and the models include the API version number, the API version is explicit as well.

```
__deserialize_model({ "name": "source" }, V1ConfigMapEnvSource)
```

**Comment by chekolyn** [comment]
@davlum i think a klaas/response_type str parameter in here https://github.com/chekolyn/python/blob/json_deserialize/kubernetes/utils/deserialize.py#L238 would do pass the klass to the __deserialize_model via this line: https://github.com/kubernetes-client/python/blob/ca4f31198e9748f1075af0e02abb8f7e1e174159/kubernetes/client/api_client.py#L289

I will amend/add this  tonight. This is still work in progress and I thank you for the suggestion.

**Comment by remram44** [comment]
Passing in the `kind` is acceptable for my use case, I always know what kind of object I am trying to deserialize.

**Comment by davlum** [comment]
Any mouvement on this?

**Comment by marcellodesales** [comment]
I Can't believe there's no way to deserialize to and from yaml

**Comment by marcellodesales** [comment]
Alright, I saw that https://github.com/kubernetes-client/python/issues/574#issuecomment-405400414 works... It's awkward, but works...

**Comment by marcellodesales** [comment]
@davlum try the method at https://github.com/kubernetes-client/python/issues/574#issuecomment-405400414 works... I needed to clone objects and change the names and properties. So I had it transformed into a `dict`, worked on that lavel, and then converted to `yaml`... That's how far I got...

**Comment by davlum** [comment]
That's not quite what I'm looking for. I would like to go from JSON/YAML to client objects defined within this library. If I'm not mistaken your example returns JSON from the client, modifies it and writes back YAML, never making the conversion from YAML -> client object.

**Comment by roycaihw** [comment]
> the deserialize() method that exists expects a requests.Response, not JSON

Reading [deserialize](https://github.com/kubernetes-client/python/blob/390f3cfd2bd4099f7e17b7196a37b201f291d62f/kubernetes/client/api_client.py#L236-L256), it mostly needs the `data` from a `RESTResponse`. The method could be reused for JSON by mocking a `RESTResponse` object, like what @chekolyn did in https://github.com/kubernetes-client/python/pull/989/files#diff-0204f9aa3f2d80c3fe32d62f37882dfaR256-R263

**Comment by piroszhog** [comment]
That's how I have to do it now. Will be glad to remove this hack after this issue close.

```
class FakeKubeResponse:
    def __init__(self, obj):
        import json
        self.data = json.dumps(obj)

fake_kube_response = FakeKubeResponse(initial_object_json)
v1pod = api_client.deserialize(fake_kube_response, 'V1Pod')
```

**Comment by roycaihw** [comment]
@piroszhog I think that is the right workaround at the moment, and this issue should be closed once #989 is finished.

**Comment by postmaxin** [comment]
> That's how I have to do it now:
> 
> ```python
> class FakeKubeResponse:
>     def __init__(self, obj):
>         import json
>         self.data = json.dumps(obj)
> 
> fake_kube_response = FakeKubeResponse(initial_object_json)
> v1pod = api_client.deserialize(fake_kube_response, 'V1Pod')
> ```
> 
> Will be glad to remove this hack after this issue close.

To make this hack work for me, I had to add explicit `datetime` deserialization. I'm not sure but I think this may be because I'm also doing funky stuff like deepcopies in my code;

```
[...]
            def _time_dumps(val):
                if isinstance(val, datetime.datetime):
                    return iso8601time(val)
                else:
                    raise TypeError(
                        "Don't know how to dump {}".format(type(val)))
            self.data = json.dumps(body, default=_time_dumps)
[...]
```

**Comment by piroszhog** [comment]
@postmaxin I never got such kind of problems. Maybe dates in your initial json are not in RFC 3339 format, which Kubernetes requires https://stackoverflow.com/a/522281

**Comment by cmoulliard** [comment]
Will this issue be implemented to easily convert Json to K8s objects ?

**Comment by klarose** [comment]
Any more interest in this?

**Comment by remram44** [comment]
I don't have enough interest to fight the bot ad-vitam. Feel free to close it like you did #340 and pretend it's not a problem. It's not like you own anyone anything, even civility.

Bye!

**Comment by aantn** [comment]
If anyone is still struggling with this, you can do it with [Hikaru](https://hikaru.readthedocs.io/). I'm not affiliated with the project in any way, but I've used it in a [my own](https://github.com/robusta-dev/robusta) open source projects to convert YAML/Json to Python objects and back.

**Comment by pigletfly** [comment]
the origin PR https://github.com/kubernetes-client/python/pull/989 is closed, so can we reopen this issue?

**Comment by pigletfly** [comment]
@roycaihw

**Comment by D0rkKnight** [comment]
Hi, I'm interested in contributing to this issue. How can I get started?

**Comment by remram44** [comment]
The project might be rotten but the issue is fresh

/remove-lifecycle stale

**Comment by remram44** [comment]
Sorry if I didn't use the right incantation to scare away the bots

/remove-lifecycle rotten

**Comment by msw-kialo** [comment]
This is still a desired feature to e.g. make testing simpler.

/remove-lifecycle rotten
